src/zhinst/qcodes/qcodes_adaptions.py

```diff
@@ -197,27 +197,42 @@ def __init__(
         self.get_raw = kwargs["get_cmd"]
         self.set_raw = kwargs["set_cmd"]
         self.get = self._wrap_get(self.get_raw)
-        self.set = self._wrap_set(self.set_raw)
+        self._get_set_return = self._wrap_get(lambda: self._set_return)
+        self.set = self._wrap_set(self.set_wrapper)
         self._snapshot_cache = snapshot_cache
         self._zi_node = zi_node
         self._tk_node = tk_node
+        self._set_return: t.Any = None
 
     def __call__(self, *args, **kwargs):
         if len(args) == 0:
             if self.gettable:
                 return self.get(**kwargs)
-            else:
-                raise NotImplementedError(
-                    "no get cmd found in" + f" Parameter {self.name}"
-                )
-        else:
-            if self.settable:
-                self.set(*args, **kwargs)
-                return None
-            else:
-                raise NotImplementedError(
-                    "no set cmd found in" + f" Parameter {self.name}"
-                )
+            raise NotImplementedError(
+                "no get cmd found in" + f" Parameter {self.name}"
+            )
+        if self.settable:
+            self.set(*args, **kwargs)
+            return self._get_set_return() if self._set_return is not None else None
+        raise NotImplementedError(
+            "no set cmd found in" + f" Parameter {self.name}"
+        )
+
+    def set_wrapper(self, *args, **kwargs) -> None:
+        """Wrapper for the set command that stored the returned value.
+        
+        QCoDeS does not provide a way to return a value for the set command.
+        However a `deep` command in LabOne does return the acknowledged value
+        by the device. This function in combination with the custom `__call__`
+        method bypasses this problem.
+
+        This function simply forwards everything to the set_raw (toolkit set
+        command) and buffers the result.
+
+        Warning: Although the `_set_return` variable is persistent after the
+        call it is not indented to be used apart from the `__call__ method,
+        """
+        self._set_return = self.set_raw(*args, **kwargs)
```

Comment:
Am I misunderstanding something ? It looks to me like calling `thisparameter(new_value)` will call `self._get_set_return` which will effectily do anther get but calling `thisparameter.set(value)` will call get again? 

\

Reply:
There are two things going on here. LabOne has two different set types sync and async. Async is the default and does not block but also does not return anything (obviously since it does not wait for the device response). Sync set blocks until the value was acknowledged by the device and returns the acknowledged value. A sync set is indicated by a `deep=True` keyword in the set command. 

So  the set command (`self.set_raw(*args, **kwargs)`) only returns a value if `deep=True`. In that case the proposed changes in line 216 will call `self._get_set_return`. But `self._get_set_return` will not do any call to the device QCoDeS will call a lambda that just returns the buffered returned value for sync set command.

I am happy for another suggestion ... but I wanted to avoid re implementing QCoDeS stuff again 😁

Reply:
Thanks for the explanation.  My question was more as far as I can see
```
instr.myparameter(newvalue, deep=True)
```
will return the value and update it

but 

```
instr.myparameter.set(value, deep=True) 

```
will not since calling set will not call `__call__` on the parameter so never touch the code on line 216 as far as I can see.

(I am also not really a fan of there being 2 ways to call set on a parameter but that has been the case for a long time so hard to change)

Reply:
good point ... I did not consider this the option of using `.set` as an end user. 

I think I need to reimplement the fix again to also work for the `.set` command